Isolate non gradle related code to a common ruler-cli module as part of the Bazel migration

ViktorPetrovski opened the pull request:
### What has changed
As a first step for the migration, we want to isolate code that is heavily dependant on Gradle dependencies. The idea is to create a ruler-cli that will be built as a standalone jar and will get the required information as command-line arguments.

* In this PR step ruler-cli is added as a module dependent on ruler-gradle-plugin just for simplicity in this first step.

* The code changes in this PR a subject to change and is not final

### Why was it changed
As we are moving away from Gradle in Spotify we want to support building ruler as a grade plugin and as a bazel plugin.

ViktorPetrovski commented on buildSrc/src/main/kotlin/Dependencies.kt:
```diff
@@ -57,9 +57,9 @@ object Dependencies {
 
     object Versions {
-        const val ANDROID_GRADLE_PLUGIN = "7.2.1" // https://mvnrepository.com/artifact/com.android.tools.build/gradle?repo=google
+        const val ANDROID_GRADLE_PLUGIN = "7.4.1" // https://mvnrepository.com/artifact/com.android.tools.build/gradle?repo=google
```
Only changed because of local problems, not part of the migration.

ViktorPetrovski commented on ruler-cli/src/main/java/com/spotify/ruler_cli/Main.kt:
```diff
@@ -0,0 +1,5 @@
+package com.spotify.ruler_cli
+
+class Main {
+    // TODO: Implement argument parsing logic.
```
A place where the main method will parse the cli arguments and distribute them accordingly

ViktorPetrovski commented on ruler-gradle-plugin/src/main/kotlin/com/spotify/ruler/plugin/dependency/EntryParser.kt:
```diff
@@ -0,0 +1,85 @@
+import java.io.File
+
+/** Responsible for parsing and extracting entries from dependencies. */
+class EntryParser {
```
Wouldn't keep this class at all, thinking of moving the logic inside the `RulerTask` class

ViktorPetrovski commented on sample/app/src/main/proguard-rules.pro:
```diff
@@ -0,0 +1,2 @@
+-keep class org.jetbrains.annotations.** { *; }
+-dontwarn org.jetbrains.annotations.**
```
This file is not needed can be removed, I've added it to test pro guard for bazel

nathan3d replied:
deleted

ViktorPetrovski commented on ruler-cli/src/main/java/com/spotify/ruler/cli/RulerCli.kt:
```diff
@@ -0,0 +1,118 @@
+
+    override fun provideResourceMappingFile(): File? {
+        return null
+        TODO("Not yet implemented")
```
We should add the resource mapping file as an optional argument

nathan3d replied:
done

ViktorPetrovski commented on ruler-cli/src/main/java/com/spotify/ruler/cli/RulerCli.kt:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    override fun provideDependencies(): Map<String, List<DependencyComponent>> {
+        val json = Json.decodeFromStream<ModuleMap>(dependencyMap.inputStream())
```
Shouldn't we make this lazy as well to avoid reDecoding on each call?

nathan3d replied:
fixed

ViktorPetrovski commented on ruler-common/src/main/java/com/spotify/rulercommon/BaseRulerTask.kt:
```diff
@@ -0,0 +1,117 @@
+    fun print(content: String)
+    fun provideMappingFile(): File?
+    fun provideResourceMappingFile(): File?
+    //    fun provideOmitFileBreakdown(): Boolean
```
Nit: AFAIK this was moved to Ruler config so we should remove it

nathan3d replied:
fixed

ViktorPetrovski commented on ruler-frontend-tests/src/test/kotlin/com/spotify/ruler/frontend/testutil/WebDriverExtension.kt:
```diff
@@ -47,6 +46,7 @@ class WebDriverExtension : BeforeAllCallback, BeforeEachCallback, AfterEachCallb
             .addArguments("--start-maximized")
             .addArguments("--no-sandbox")
             .addArguments("--disable-dev-shm-usage")
+            .addArguments("--remote-allow-origins=*")
```
Is this needed? 🤔

nathan3d replied:
yeah, fixes some Selenium issues

simonschiller replied:
This (and maybe the Selenium dependency upgrades) seem to be a separate issue not related to this PR (i.e. #110 is also failing). Probably should be a separate PR to unblock other PRs while this one is getting reviewed?

nathan3d replied:
done

simonschiller commented on ruler-common/src/main/java/com/spotify/rulercommon/BaseRulerTask.kt:
```diff
@@ -0,0 +1,116 @@
+
+const val FEATURE_NAME = "base"
+
+interface BaseRulerTask {
```
Instead of extending a common base class (or implementing an interface), I feel like it would be cleaner and easier to test if we'd have:
- Something like a `Variant` interface that encapsulates the variant-specific logic. This should be a pure interface without any implementation
- Specific implementations (`GradleVariant`, `CliVariant`, ...) for each variant
- A common `RulerRunner` class that takes in the variant and actually runs the common logic. I.e. something like this:
```kotlin
class RulerRunner {
    fun run(variant: Variant)
}
```
- Each entry point for a variant (`RulerTask` for Gradle, `RulerCli` for CLI, ...) would just set up the variant and pass it to the runner:
```kotlin
class RulerTask {
    fun run() {
        // Maybe some more Gradle-specific logic

        val variant = buildGradleVariant()
        rulerRunner.run(variant)

        // Maybe some more Gradle-specific logic
    }
}
```

That way the code becomes a lot easier to test (you can have a `TestVariant` for unit tests to test behaviour) and easier to expand in the future. Composition is [typically preferrable over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance), even if "multi-inheritance" with Kotlin interfaces is technically possible.

nathan3d replied:
Interesting idea, but I think will consider it for later. I think the way we did it now was a smaller change, and is reasonable for what we need to accomplish.

simonschiller commented on ruler-common/src/main/java/com/spotify/rulercommon/dependency/ArtifactParser.kt:
Unit tests for this and `DependencyParser` seem to be missing, or did I miss something?

simonschiller reviewed (commented):
Overall looks good to me. There is a lot going on in this PR, I feel like this could have been split up more (i.e. one for the `ruler-common` module, second one for `ruler-cli`, not sure why there were test changes, ...). Would have been a lot easier to understand the changes and why they were done. Especially since I'm missing the internal context.

I feel like I was a little nit-picky in some areas. Should go without saying, but of course feel free to ignore my comments, as I'm no longer an owner of Ruler. I just wanted to take the opportunity to share context where I can.

nathan3d commented on ruler-common/src/main/java/com/spotify/rulercommon/BaseRulerTask.kt:
```diff
@@ -0,0 +1,116 @@
+* limitations under the License.
+*/
+
+package com.spotify.rulercommon
```
Updated

nathan3d commented on ruler-e2e-tests/src/test/kotlin/com/spotify/ruler/e2e/ReleaseReportTest.kt:
```diff
@@ -43,14 +43,15 @@ class ReleaseReportTest {
         assertThat(app.owner).isEqualTo("default-team")
 
-        assertThat(app.files).comparingElementsUsing(Correspondence.file()).containsExactly(
+        assertThat(app.files).comparingElementsUsing(Correspondence.file()).containsAtLeastElementsIn(
```
Reverted

nathan3d commented on ruler-common/src/main/java/com/spotify/rulercommon/sanitizer/ClassNameSanitizer.kt:
```diff
@@ -33,6 +33,7 @@ class ClassNameSanitizer {
             .removeSurrounding("L", ";") // La/b/C; -> a/b/C
             .removeSuffix(".class") // a/b/C.class -> a/b/C
             .replace('/', '.') // a/b/c -> a.b.C
+            .replace("/lib/res/","/res/")
```
done

nathan3d commented on ruler-gradle-plugin/bin/main/com/spotify/ruler/plugin/RulerPlugin.kt:
Not sure what happened here. Maybe some generated files? Anyway... removed

ViktorPetrovski commented:
@simonschiller Thanks for the review, very much appreciated! I think the main reason for having one PR is that we were very uncertain where we will end up when we started this, this was mostly done through a lot of pairing sessions so yeah I can understand it was a bit overwhelming for you to get the full idea just by looking at the PR for us easier since we were doing this together. We addressed most of the comments and added more e2e tests for the RulerCli module.

ViktorPetrovski commented:
@simonschiller We are going to go ahead and merge the PR so we can continue with the Spotify Extension plugin integration, but feel free to add any comment and we can address it as separate PR!
